pageserver/src/http/routes.rs

```diff
@@ -2353,6 +2357,84 @@ async fn get_utilization(
         .map_err(ApiError::InternalServerError)
 }
 
+/// Report on the largest tenants on this pageserver, for the storage controller to identify
+/// candidates for splitting
+async fn post_top_n_tenants(
+    mut r: Request<Body>,
+    _cancel: CancellationToken,
+) -> Result<Response<Body>, ApiError> {
+    check_permission(&r, None)?;
+    let request: TopNTenantShardsRequest = json_request(&mut r).await?;
+    let state = get_state(&r);
+
+    fn get_size_metric(sizes: &TopNTenantShardItem, order_by: &TopNSorting) -> u64 {
+        match order_by {
+            TopNSorting::ResidentSize => sizes.resident_size,
+            TopNSorting::MaxLogicalSize => sizes.max_logical_size,
+        }
+    }
+
+    let mut top_n = Vec::new();
+
+    // FIXME: this is a lot of clones to take this tenant list
+    for (tenant_shard_id, tenant_slot) in state.tenant_manager.list() {
+        if let Some(shards_lt) = request.where_shards_lt {
+            // Ignore tenants which already have >= this many shards
+            if tenant_shard_id.shard_count >= shards_lt {
+                continue;
+            }
+        }
+
+        let sizes = match tenant_slot {
+            TenantSlot::Attached(tenant) => tenant.get_sizes(),
+            TenantSlot::Secondary(_) | TenantSlot::InProgress(_) => {
+                continue;
+            }
+        };
+        let metric = get_size_metric(&sizes, &request.order_by);
+
+        if let Some(gt) = request.where_gt {
+            // Ignore tenants whose metric is <= the lower size threshold, to do less sorting work
+            if metric <= gt {
+                continue;
+            }
+        };
+
+        // Insert to vector keeping it sorted
+        match top_n.last() {
+            None => {
+                // Top N list is empty: candidate becomes first member
+                top_n.push((metric, sizes));
+            }
+            Some(last) if last.0 > metric && top_n.len() < request.limit => {
+                // Lowest item in list is greater than our candidate, but we aren't at limit yet: push to end
+                top_n.push((metric, sizes));
+            }
+            Some(last) if last.0 > metric => {
+                // List is at limit and lowest value is greater than our candidate, drop it.
+            }
+            Some(_) => {
+                // Find location to insert
+                for i in 0..top_n.len() {
+                    if top_n[i].0 < metric {
```

Comment:
Maybe use `[Vec::binary_search](https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.binary_search)` instead and then match on the `Err` case?

Reply:
It's better to use BinaryHeap for finding largest tenants:
```rust
let mut heap = BinaryHeap::with_capacity(request.limit);

for ... {
  ...
  heap.push(Reverse((metric, sizes)));

  if heap.len() > request.limit {
    heap.pop();
  }
}
```

Reply:
that's even better, yeah

Reply:
The flat vec is probably optimal for the way we use this API from the controller (with a limit of ~10), but it's fair to point out that people could call the API for higher counts, so adopting binary heap here: df8d888d90493b34d604f175fe928f2a190d721c